tests/qit/test-package/tests/woopayments/merchant/merchant-orders-refund-failures.spec.ts

```diff
@@ -0,0 +1,125 @@
+	[
+		'refund amount in line item',
+		'greater than maximum',
+		selectorLineAmount,
+		'100',
+	],
+	[ 'refund amount in line item', 'negative', selectorLineAmount, '-1' ],
+	[
+		'total refund amount',
+		'greater than maximum',
+		selectorTotalAmount,
+		'100',
+	],
+	[ 'total refund amount', 'negative', selectorTotalAmount, '-1' ],
+];
+let orderId: string;
+
+test.describe( 'Order > Refund Failure', { tag: '@merchant' }, () => {
+	test.beforeAll( async ( { customerPage, adminPage } ) => {
+		// Place an order to refund later
+		await emptyCart( customerPage );
+		orderId = await placeOrderWithCurrency( customerPage, 'USD' );
+		await ensureOrderIsProcessed( adminPage, orderId );
+	} );
+
+	dataTable.forEach( ( [ fieldName, valueDescription, selector, value ] ) => {
+		test.describe( 'Invalid ' + fieldName, () => {
+			test.beforeEach( async ( { adminPage } ) => {
+				// Open the order
+				await goToOrder( adminPage, orderId );
+
+				// Sometimes the element is not clickable due to the header getting on the way. This seems to
+				// only happen in CI for WC 7.7.0 so the workaround is to remove those elements.
+				const hideElementIfExists = ( selectorToHide: string ) => {
+					const element = document.querySelector( selectorToHide );
+					if ( element ) {
+						element.outerHTML = '';
+					}
+				};
+				await adminPage.evaluate(
+					hideElementIfExists,
+					'.woocommerce-layout__header'
+				);
+				await adminPage.evaluate( hideElementIfExists, '#wpadminbar' );
+
+				// Click the Refund button
+				const refundItemsButton = adminPage
+					.getByRole( 'button', {
+						name: 'Refund',
+					} )
+					.first();
+				await refundItemsButton.click();
+
+				// Verify the refund section shows
+				await adminPage.waitForSelector( 'div.wc-order-refund-items' );
+
+				// Verify Refund via WooPayments button is displayed
+				await adminPage.waitForSelector( 'button.do-api-refund' );
+			} );
+
+			test(
+				`should fail refund attempt when ${ fieldName } is ${ valueDescription }`,
+				{ tag: '@critical' },
+				async ( { adminPage } ) => {
+					// Initiate refund attempt
+					await adminPage.locator( selector ).first().fill( value );
+
+					const refundButton = await adminPage.waitForSelector(
+						'.do-api-refund',
+						{
+							state: 'visible',
+						}
+					);
+					const refundButtonText: string =
+						await refundButton.textContent();
+					expect( refundButtonText ).toMatch(
+						/Refund .* via WooPayments.+/
+					);
+
+					function* dialogHandler( dialog: Dialog ) {
+						yield dialog.accept();
+						expect( dialog.message() ).toBe(
+							'Invalid refund amount'
+						);
+						yield dialog.accept();
+					}
+
+					// Confirm the refund. There will be two dialogs shown, one for the refund
+					// confirmation, one for the error message. Accept the first one, and
+					// verify the second one.
+					adminPage.on(
+						'dialog',
+						( dialog ) => dialogHandler( dialog ).next().value
+					);
```

Comment:
TBH, this took me a while to understand 😆 I'm not used to seeing generator patterns - it's kinda cool.